[jit] make test context managers exception safe

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22504 [WIP] classtypes hold weak_ptr
* #22221 [jit] _script_compile and _script_class_compile add to the python CU
* #22207 [jit] refactor self to be a class again
* #22206 [jit] Give functions qualified names
* #22314 [jit] use save/load for emitFunctionHook
* #22205 [jit] remove unused argument in import.cpp
* #22204 [jit] improvements to QualifiedName
* #22203 [jit] refactoring of module/object
* #22202 [jit] Make CompilationUnit own Functions
* **#22502 [jit] make test context managers exception safe**

Context managers are not exception safe, so if an exception is thrown in
the `with` block, the part after `yield` will not be called. Which leads
to very weird results if you have a test failure and it doesn't clean up
after itself.

Differential Revision: [D16121783](https://our.internmc.facebook.com/intern/diff/D16121783)

suo committed:
[jit] make test context managers exception safe

Context managers are not exception safe, so if an exception is thrown in
the `with` block, the part after `yield` will not be called. Which leads
to very weird results if you have a test failure and it doesn't clean up
after itself.

bwasti reviewed (approved):
lgtm, basically a custom version of `contextlib.closing`

suo committed:
Update on "[jit] make test context managers exception safe"

[jit] make test context managers exception safe

Context managers are not exception safe, so if an exception is thrown in
the `with` block, the part after `yield` will not be called. Which leads
to very weird results if you have a test failure and it doesn't clean up
after itself.

gh-metadata: pytorch pytorch 22502 gh/suo/75/head
